src/ophyd_async/core/device.py

```diff
@@ -83,12 +91,25 @@ async def connect(self, mock: bool = False, timeout: float = DEFAULT_TIMEOUT):
         timeout:
             Time to wait before failing with a TimeoutError.
         """
-        coros = {
-            name: child_device.connect(mock=mock, timeout=timeout)
-            for name, child_device in self.children()
-        }
-        if coros:
-            await wait_for_connection(**coros)
+
+        previous_connect_ok = (
+            self._connect_task
+            and self._connect_task.done()
+            and not self._connect_task.exception()
+        )
```

Comment:
Please can we support switching between `mock=True` and `mock=False`.

Something like:
```suggestion
            and self._connect_mock_arg == mock
        )
```